engine/access/rpc/backend/backend.go

```diff
@@ -539,34 +539,82 @@ func findAllExecutionNodes(
 	allENs, err := state.Final().Identities(filter.HasRole[flow.Identity](flow.RoleExecution))
 	if err != nil {
-		return nil, fmt.Errorf("failed to retreive all execution IDs: %w", err)
+		return nil, fmt.Errorf("failed to retrieve all execution IDs: %w", err)
 	}
 
-	// first try and choose from the preferred EN IDs
+	// choose from preferred EN IDs
 	var chosenIDs flow.IdentityList
 	if len(preferredENIdentifiers) > 0 {
-		// find the preferred execution node IDs which have executed the transaction
-		chosenIDs = allENs.Filter(filter.And(filter.HasNodeID[flow.Identity](preferredENIdentifiers...),
-			filter.HasNodeID[flow.Identity](executorIDs...)))
-		if len(chosenIDs) > 0 {
-			return chosenIDs.ToSkeleton(), nil
-		}
+		chosenIDs := chooseFromPreferredENIDs(allENs, executorIDs)
+		return chosenIDs.ToSkeleton(), nil
 	}
 
 	// if no preferred EN ID is found, then choose from the fixed EN IDs
 	if len(fixedENIdentifiers) > 0 {
 		// choose fixed ENs which have executed the transaction
 		chosenIDs = allENs.Filter(filter.And(
 			filter.HasNodeID[flow.Identity](fixedENIdentifiers...),
-			filter.HasNodeID[flow.Identity](executorIDs...)))
+			filter.HasNodeID[flow.Identity](executorIDs...),
+		))
 		if len(chosenIDs) > 0 {
 			return chosenIDs.ToSkeleton(), nil
 		}
-		// if no such ENs are found then just choose all fixed ENs
+		// if no such ENs are found, then just choose all fixed ENs
 		chosenIDs = allENs.Filter(filter.HasNodeID[flow.Identity](fixedENIdentifiers...))
 		return chosenIDs.ToSkeleton(), nil
 	}
 
-	// If no preferred or fixed ENs have been specified, then return all executor IDs i.e. no preference at all
+	// if no preferred or fixed ENs have been specified, then return all executor IDs i.e., no preference at all
 	return allENs.Filter(filter.HasNodeID[flow.Identity](executorIDs...)).ToSkeleton(), nil
 }
+
+// chooseFromPreferredENIDs finds the subset of execution nodes if preferred execution nodes are defined.
+// If preferredENIdentifiers is set and there are less than maxNodesCnt nodes selected, than the list is padded up to
+// maxNodesCnt nodes using the following order:
+// 1. Use any EN with a receipt.
+// 2. Use any preferred node not already selected.
+// 3. Use any EN not already selected.
+func chooseFromPreferredENIDs(allENs flow.IdentityList, executorIDs flow.IdentifierList) flow.IdentityList {
+	var chosenIDs flow.IdentityList
+
+	// filter for both preferred and executor IDs
+	chosenIDs = allENs.Filter(filter.And(
+		filter.HasNodeID[flow.Identity](preferredENIdentifiers...),
+		filter.HasNodeID[flow.Identity](executorIDs...),
+	))
+
+	// function to add nodes to chosenIDs if they are not already included
+	addIfNotExists := func(candidates flow.IdentityList) {
+		for _, en := range candidates {
+			_, exists := chosenIDs.ByNodeID(en.NodeID)
+			if !exists && len(chosenIDs) < maxNodesCnt {
+				chosenIDs = append(chosenIDs, en)
+				if len(chosenIDs) >= maxNodesCnt {
+					return
+				}
+			}
```

Comment:
we can leave it to the caller to only call if there is room to add at least one
```suggestion
			_, exists := chosenIDs.ByNodeID(en.NodeID)
			if !exists {
				chosenIDs = append(chosenIDs, en)
				if len(chosenIDs) >= maxNodesCnt {
					return
				}
			}
```